src/file_pipe_log/pipe_builder.rs

```diff
@@ -100,141 +103,142 @@ impl<F: FileSystem> DualPipesBuilder<F> {
-            fs::create_dir(root_path)?;
-            self.dir_lock = Some(lock_dir(root_path)?);
-            return Ok(());
-        }
-        if !root_path.is_dir() {
-            return Err(box_err!("Not directory: {}", root_path.display()));
+        self.scan_dir(self.cfg.dir.clone())?; // scan the main dir
+        if let Some(dir) = self.cfg.spill_dir.clone() {
+            self.scan_dir(dir)?; // scan the auxiliary dir
         }
-        self.dir_lock = Some(lock_dir(root_path)?);
+        // Sorts the expected `file_list` according to `file_seq`.
+        self.append_files.sort_by(|a, b| a.seq.cmp(&b.seq));
+        self.rewrite_files.sort_by(|a, b| a.seq.cmp(&b.seq));
+        self.recycled_files.sort_by(|a, b| a.seq.cmp(&b.seq));
 
-        let (mut min_append_id, mut max_append_id) = (u64::MAX, 0);
-        let (mut min_rewrite_id, mut max_rewrite_id) = (u64::MAX, 0);
-        let (mut min_recycled_id, mut max_recycled_id) = (u64::MAX, 0);
-        fs::read_dir(root_path)?.for_each(|e| {
-            if let Ok(e) = e {
-                let p = e.path();
-                if !p.is_file() {
-                    return;
-                }
-                let name = p.file_name().unwrap().to_str().unwrap();
-                match FileId::parse_file_name(name) {
-                    Some(FileId {
-                        queue: LogQueue::Append,
-                        seq,
-                    }) => {
-                        min_append_id = cmp::min(min_append_id, seq);
-                        max_append_id = cmp::max(max_append_id, seq);
-                    }
-                    Some(FileId {
-                        queue: LogQueue::Rewrite,
-                        seq,
-                    }) => {
-                        min_rewrite_id = cmp::min(min_rewrite_id, seq);
-                        max_rewrite_id = cmp::max(max_rewrite_id, seq);
-                    }
-                    _ => {
-                        if let Some(seq) = parse_recycled_file_name(name) {
-                            min_recycled_id = cmp::min(min_recycled_id, seq);
-                            max_recycled_id = cmp::max(max_recycled_id, seq);
-                        }
-                    }
+        // Validate and clear obsolete metadata and log files.
+        for (queue, files, is_recycled_file) in [
+            (LogQueue::Append, &mut self.append_files, false),
+            (LogQueue::Rewrite, &mut self.rewrite_files, false),
+            (LogQueue::Append, &mut self.recycled_files, true),
+        ] {
+            // Check the file_list and remove the hole of files.
+            let mut invalid_idx = 0_usize;
+            for (i, file_pair) in files.windows(2).enumerate() {
+                // If there exists a black hole or duplicate scenario on FileSeq, these
+                // files should be skipped and cleared.
+                if file_pair[1].seq - file_pair[0].seq != 1 {
```

Comment:
is duplicate file tested?

Reply:
Yep, this case has already been supplemented in `src/engine.rs +2735` as following shows:
```Rust
// Case 4: abnormal case - duplicate FileSeq among different dirs.
{
    // Prerequisite: choose several files and duplicate them to main dir.
    let mut file_count = 0;
    std::fs::read_dir(paths[1]).unwrap().for_each(|e| {
        let p = e.unwrap().path();
        let file_name = p.file_name().unwrap().to_str().unwrap();
        if let Some(FileId {
            queue: LogQueue::Append,
            seq: _,
        }) = FileId::parse_file_name(file_name)
        {
            let mut dst_path = PathBuf::from(&paths[0]);
            dst_path.push(file_name);
            if file_count % 2 == 0 {
                std::fs::copy(p, dst_path).unwrap();
            }
            file_count += 1;
        }
    });
}
let engine = engine.reopen();
// Duplicate log files will be skipped and cleared.
assert!(engine.file_span(LogQueue::Append).0 > start_5);
```